src/reducers/moveThought.ts

```diff
@@ -167,147 +41,11 @@ const moveThought = (state: State, { oldPath, newPath, offset, skipRerank, newRa
-      const sourceThoughtlexeme = updatedThoughtIndex[hashedThought]!
-
-      // remove source thought from the lexeme entry as it's context index entry will be deleted
-      const updatedLexeme: Lexeme = {
-        ...sourceThoughtlexeme,
-        contexts: sourceThoughtlexeme!.contexts.filter(thoughtId => thoughtId !== sourceThought.id),
-      }
-
-      const finalUpdatedThoughtIndex = {
-        ...updates.thoughtIndexUpdates,
-        [hashedThought]: updatedLexeme,
-      }
-
-      const mergeUpdates = {
-        ...updates,
-        contextIndexUpdates: {
-          ...updates.contextIndexUpdates,
-          // after children are moved to the duplicate context, the source parent must be deleted
-          [sourceThought.id]: null,
-        },
-        thoughtIndexUpdates: finalUpdatedThoughtIndex,
-      }
-
-      return mergeUpdates
-    }
-
-    const mergeUpdates = !!duplicateThought && mergeChildren()
-
-    /** Remove source thought from the parent. */
-    const updatedSourceParentChildren = () =>
-      getAllChildrenByContextHash(state, sourceParentThought.id).filter(child => child !== movingThoughtId)
-
-    const finalContextIndexUpdates = {
-      ...(!sameContext
-        ? {
-            [sourceParentThought.id]: {
-              ...sourceParentThought,
-              children: updatedSourceParentChildren(),
-              lastUpdated: timestamp(),
-              updatedBy: getSessionId(),
-            },
-          }
-        : {}),
-      ...(mergeUpdates
-        ? mergeUpdates.contextIndexUpdates
-        : {
-            // Only update source thought and move it to the desination if there is no duplicate merge. Because source thought would be deleted if merged.
-            [sourceThought.id]: {
-              ...sourceThought,
-              parentId: destinationThought.id,
-              rank,
-              archived,
-              lastUpdated: timestamp(),
-              updatedBy: getSessionId(),
-            },
-            ...(!sameContext
-              ? {
-                  [destinationThought.id]: {
-                    ...destinationThought,
-                    children: [...destinationThought.children, sourceThought.id],
-                    lastUpdated: timestamp(),
-                    updatedBy: getSessionId(),
-                  },
-                }
-              : {}),
-          }),
-    }
-
-    const finalUpdates: RecursiveMoveData = {
-      contextIndexUpdates: finalContextIndexUpdates,
-      thoughtIndexUpdates: mergeUpdates ? mergeUpdates.thoughtIndexUpdates : {},
-      pendingMerges: mergeUpdates ? mergeUpdates.pendingMerges : [],
-    }
-
-    return finalUpdates
-  }
-
-  const { thoughtIndexUpdates, contextIndexUpdates } = recursiveMove(movingThoughtId, newParentId, newRank, state)
+  const { thoughtIndexUpdates, contextIndexUpdates, pendingMerges } = move(state, {
+    sourceThoughtPath: oldPath,
+    destinationThoughtPath: rootedParentOf(state, newPath),
+    rank: newRank,
+  })
```

Comment:
I'd like to discuss some architectural considerations with the new organization of logic.

What is the difference between `moveThought` and `move`? How are responsibilities divided? Does this result in proper encapsulation and an easy-to-understand low-level api? Compare with `addThought`, `editThought`, and `removeThought` which make up a clear CRUD interface.

At first glance, it looks like all functionality has moved to `move` except the reranking logic. Why not put all the logic in `moveThought` and use that everywhere? Could slightly different behaviors be achieved with an option? Which results in more complexity, the introduction of an option, or splitting out the move logic?

I look forward to hearing your thoughts. Thanks!

Reply:
I was meaning to discuss this with you. The reason I separated `move` and `mergeThoughts` are as follows:


`move` moves thought to a destination and handles duplicate merging if found any. It uses `mergeThoughts` in case of duplicate thoughts. It returns thought indices updates and pending merges are found any.

`mergeThoughts` merges two given thoughts with the same value. It returns the same thing as the `move`. It uses `move` but is the logic is specialized to merge two thoughts.

These two functions are dependent on each other and help create two distinct reducers. `moveThought` and `mergePending`. I agree that they are totally not meant to be selectors. 

I just wanted a way to create these two reducers `moveThought` and `mergePending` without having to duplicate this logic since they are fundamentally dependent on the move and merge logic. However `moveThought` first tries to move the thought and only merges when needed and `mergePending` directly tries to merge the given thoughts and uses `move` logic to achieve that. I found myself spending too much time deciding how to architecture it. So instead I decided to first get it working. That was the original reason for the separation of these `move` and `mergeThought`. I just kept them in `selectors` for the time being. 😅

Reply:
Now that I can think of we can have `move` and `mergeThought` logic as part of the main `moveThought` reducer and then export `mergThought` logic from that file and use it in `mergePending` reducer. What do you think of that ?